[ARM] Record store with pre/post-indexed addressing as `mayStore`

antoniofrighetto opened the pull request:
A miscompilation issue observed during machine sinking has been addressed with improved handling.

Fixes: https://github.com/llvm/llvm-project/issues/121299.

davemgreen commented:
Thanks for the fix for this. Does UnmodeledSideEffects not imply MayStore? There are a number of other store instructions (for various architectures) that have UnmodeledSideEffects but not MayStore.

It looks like there is an mca test that needs updating too.

antoniofrighetto commented:
Updated, thanks! That doesn't look like to be the case for ARM, as UnmodeledSideEffects (setting `hasSideEffects = 1` only) doesn't seem to imply MayStore (I'd personally set it explicitly anyways if it were the case though, for consistency).

davemgreen reviewed (approved):
This seems like an obvious improvement on it's own. I was expecting UnmodeledSideEffects to act as a load barrier, as in MachineInstr::isLoadFoldBarrier. You could argue there is another problem in MachineSink, but providing we reliably mark store as MayStore it should be OK. I'll look to see if there are other instructions in Arm that could do with the same.

LGTM, thanks.

antoniofrighetto committed:
[ARM] Introduce test for PR121565 (NFC)

antoniofrighetto committed:
[ARM] Record store with pre/post-indexed addressing as `mayStore`

A miscompilation issue observed during machine sinking has been
addressed with improved handling.

Fixes: https://github.com/llvm/llvm-project/issues/121299.
